source/common/upstream/cluster_manager_impl.cc

```diff
@@ -1132,7 +1133,120 @@ void ClusterManagerImpl::ThreadLocalClusterManagerImpl::ClusterEntry::drainConnP
 ClusterUpdateCallbacksHandlePtr
 ClusterManagerImpl::addThreadLocalClusterUpdateCallbacks(ClusterUpdateCallbacks& cb) {
   ThreadLocalClusterManagerImpl& cluster_manager = *tls_;
-  return std::make_unique<ClusterUpdateCallbacksHandleImpl>(cb, cluster_manager.update_callbacks_);
+  return cluster_manager.addClusterUpdateCallbacks(cb);
+}
+
+OdCdsApiHandlePtr
+ClusterManagerImpl::allocateOdCdsApi(const envoy::config::core::v3::ConfigSource& odcds_config,
+                                     OptRef<xds::core::v3::ResourceLocator> odcds_resources_locator,
+                                     ProtobufMessage::ValidationVisitor& validation_visitor) {
+  // TODO(krnowak): Instead of creating a new handle every time, store the handles internally and
+  // return an already existing one if the config or locator matches. Note that this may need a way
+  // to clean up the unused handles, so we can close the unnecessary connections.
+  auto odcds = OdCdsApiImpl::create(odcds_config, odcds_resources_locator, *this, *this, stats_,
+                                    validation_visitor);
+  return OdCdsApiHandleImpl::create(*this, std::move(odcds));
+}
+
+ClusterDiscoveryCallbackHandlePtr
+ClusterManagerImpl::requestOnDemandClusterDiscovery(OdCdsApiSharedPtr odcds, std::string name,
+                                                    ClusterDiscoveryCallbackPtr callback,
+                                                    std::chrono::milliseconds timeout) {
+  ThreadLocalClusterManagerImpl& cluster_manager = *tls_;
+
+  auto [handle, discovery_in_progress, invoker] =
+      cluster_manager.cdm_.addCallback(name, std::move(callback));
+  // This check will catch requests for discoveries from this thread only. If other thread requested
+  // the same discovery, we will detect it in the main thread later.
+  if (discovery_in_progress) {
+    ENVOY_LOG(debug,
+              "cm odcds: on-demand discovery for cluster {} is already in progress, something else "
+              "in thread {} has already requested it",
+              name, cluster_manager.thread_local_dispatcher_.name());
+    // This worker thread has already requested a discovery of a cluster with this name, so nothing
+    // more left to do here.
+    return std::move(handle);
+  }
+  ENVOY_LOG(
+      debug,
+      "cm odcds: forwarding the on-demand discovery request for cluster {} to the main thread",
+      name);
+  // This seems to be the first request for discovery of this cluster in this worker thread. Rest of
+  // the process may only happen in the main thread.
+  dispatcher_.post([this, odcds = std::move(odcds), timeout, name = std::move(name),
+                    invoker = std::move(invoker),
+                    &thread_local_dispatcher = cluster_manager.thread_local_dispatcher_] {
+    // Check for the cluster here too. It might have been added between the time when this closure
```

Comment:
Let me think about this case, it's been a while since I wrote it. :)

Reply:
If some cluster is removed then we check the `pending_cluster_creations_` map to see if that cluster was requested on demand. If it's so then we notify all worker threads about this fact. So I'd say that it depends on timing when the event comes in relation to adding the cluster name to discovery manager and adding the cluster name to the `pending_cluster_creations_` map. I could imagine following scenarios, where W1T is worker1 thread, W2T is worker2 thread and MT is main thread.

case 1:

W1T: register callback for cluster `foo` in thread-local discovery manager
MT: received cluster `foo` removed for some reason

Nothing happens in the reaction to the cluster `foo` removal, because `foo` is not yet in `pending_cluster_creations_`. So eventually we would add `foo` to `pending_cluster_creations_` and use ODCDS to request the cluster and the wait for discovery results.

case 2:

W1T: register callback for cluster `foo` in thread-local discovery manager
MT: register `foo` as requested by W1T in `pending_cluster_creations_` and use ODCDS to request the cluster
MT: received cluster `foo` removed

This will invoke the callback in W1T with a result that cluster is missing.

case 3:

W1T: register callback for cluster `foo` in thread-local discovery manager
MT: register `foo` as requested by W1T in `pending_cluster_creations_` and use ODCDS to request the cluster
W2T: register callback for cluster `foo` in thread-local discovery manager
MT: received cluster `foo` removed

In this case both W1T and W2T will invoke the callbacks for cluster `foo` with a result `missing`. But since the registration of `foo` in `pending_cluster_creations_` from W2T didn't happen yet (scheduled and not yet dispatched), eventually we will request the cluster `foo` again. The callback in discovery manager won't be invoked for a second time, so it looks like we shouldn't do the discovery again. But at this point we are in the main thread, but the information about callbacks are in worker thread. Trying to avoid the superfluous discovery could be a messy and unreliable thing involving mutexes and whatnot.